src/core/json/path.cc

```diff
@@ -0,0 +1,223 @@
+
+  // TODO: for some operations we need to know the type of mismatches.
+  void Traverse(absl::Span<const PathSegment> path, const JsonType& json, const Cb& callback);
+
+  unsigned matches() const {
+    return matches_;
+  }
+
+ private:
+  enum MatchStatus {
+    OUT_OF_BOUNDS,
+    MISMATCH,
+  };
+
+  bool TraverseImpl(absl::Span<const PathSegment> path, const Cb& callback);
+
+  nonstd::expected<void, MatchStatus> PerformStep(const PathSegment& segment, const JsonType& node,
+                                                  const Cb& callback);
+
+  void DoCall(const Cb& callback, const JsonType& node) {
+    ++matches_;
+    callback(node);
+  }
+
+  using AdvanceResult = nonstd::expected<const JsonType*, MatchStatus>;
+
+  struct Item {
+    const JsonType* obj;
+    variant<monostate, JsonType::const_object_iterator, JsonType::const_array_iterator> state;
+    Item(const JsonType* o) : obj(o) {
+    }
+
+    // Returns the next object to traverse
+    // or null if traverse was exhausted or the segment does not match.
+    AdvanceResult Advance(const PathSegment& segment);
+  };
+
+  unsigned matches_ = 0;
+};
+
+auto Dfs::Item::Advance(const PathSegment& segment) -> AdvanceResult {
+  if (std::holds_alternative<monostate>(state)) {
+    // init state
+    switch (segment.type()) {
+      case SegmentType::IDENTIFIER: {
+        if (obj->is_object()) {
+          auto it = obj->find(segment.identifier());
+          if (it != obj->object_range().end()) {
+            state = it;
+            return &it->value();
+          } else {
+            return nullptr;  // exhausted
+          }
+        }
+        break;
+      }
+      case SegmentType::INDEX: {
+        unsigned index = segment.index();
+        if (obj->is_array()) {
+          if (index >= obj->size()) {
+            return make_unexpected(OUT_OF_BOUNDS);  // exhausted
+          }
+          auto it = obj->array_range().cbegin() + index;
+          state = it;
+          return &*it;
+        }
+        break;
+      }
+
+      case SegmentType::WILDCARD: {
+        if (obj->is_object()) {
+          jsoncons::range rng = obj->object_range();
+          if (rng.cbegin() == rng.cend()) {
+            return nullptr;
+          }
+          auto it = rng.begin();
+          state = it;
+          return &it->value();
+        }
+
+        if (obj->is_array()) {
+          jsoncons::range rng = obj->array_range();
+          if (rng.cbegin() == rng.cend()) {
+            return nullptr;
+          }
+          state = rng.cbegin();
+          return &*rng.cbegin();
+        }
+        break;
+      }
+    }  // end switch
+
+    return make_unexpected(MISMATCH);
+  }  // end monostate
+
+  // If segment is set to sa single value, we can't advance.
+  if (segment.type() != SegmentType::WILDCARD) {
+    return nullptr;  // exhausted
+  }
+
+  // Advance for WILDCARD.
+  if (std::holds_alternative<JsonType::const_object_iterator>(state)) {
```

Comment:
every type is checked for the state, so I suggest implementing it as a visitor

Reply:
not every, we do not check for monostate.

Reply:
at the beginning of the function you check it with  if (std::holds_alternative<monostate>(state)) {